packages/discovery/src/discovery/handlers/user/StateFromEvent.ts

```diff
@@ -0,0 +1,123 @@
+import { ContractValue, EthereumAddress } from '@l2beat/shared'
+import { utils } from 'ethers'
+import * as z from 'zod'
+
+import { DiscoveryLogger } from '../../DiscoveryLogger'
+import { DiscoveryProvider } from '../../provider/DiscoveryProvider'
+import { Handler, HandlerResult } from '../Handler'
+import { getEventFragment } from '../utils/getEventFragment'
+import { toContractValue } from '../utils/toContractValue'
+
+export type StateFromEventDefinition = z.infer<typeof StateFromEventDefinition>
+export const StateFromEventDefinition = z.strictObject({
+  type: z.literal('stateFromEvent'),
+  event: z.string(),
+  eventKeys: z.array(z.string()),
+  returnAs: z.enum(['array', 'object']),
+  groupByObjectKey: z.optional(z.string()),
+  ignoreRelative: z.optional(z.boolean()),
+})
+
+function checkEventKeysPresent(eventKeys: string[], inputs: utils.ParamType[]) {
+  const inputNames = inputs.map((input) => input.name)
+
+  if (inputNames.length !== eventKeys.length) {
+    return false
+  }
+
+  for (const eventKey of eventKeys) {
+    if (!inputNames.includes(eventKey)) {
+      return false
+    }
+  }
+  return true
+}
+
+export class StateFromEvent implements Handler {
+  readonly dependencies: string[] = []
+  private readonly fragment: utils.EventFragment
+  private readonly abi: utils.Interface
+
+  constructor(
+    readonly field: string,
+    readonly definition: StateFromEventDefinition,
+    abi: string[],
+    readonly logger: DiscoveryLogger,
+  ) {
+    this.fragment = getEventFragment(definition.event, abi, (fragment) => {
+      return checkEventKeysPresent(definition.eventKeys, fragment.inputs)
+    })
+    this.abi = new utils.Interface([this.fragment])
+  }
+
+  getEvent() {
+    return this.fragment.format(utils.FormatTypes.full)
+  }
+
+  async execute(
+    provider: DiscoveryProvider,
+    address: EthereumAddress,
+    blockNumber: number,
+  ): Promise<HandlerResult> {
+    this.logger.logExecution(this.field, ['Querying ', this.fragment.name])
+    const logs = await provider.getLogs(
+      address,
+      [this.abi.getEventTopic(this.fragment)],
+      0,
+      blockNumber,
+    )
+
+    const values = new Set<ContractValue>()
+    for (const log of logs) {
+      const parsed = this.abi.parseLog(log)
+      const value = this.definition.eventKeys
```

Comment:
Because eventKeys will now be used to filter you can get all event keys by going through `this.fragment.inputs`